man/mkify.1
.\"
.\" Copyright (c) 2008-2023 Julien Nadeau Carriere <vedge@csoft.net>
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
.\" WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
.\" INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
.\" (INCLUDING BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
.\" SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
.\" STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
.\" IN ANY WAY OUT OF THE USE OF THIS SOFTWARE EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd March 1, 2024
.Dt MKIFY 1
.Os BSDBuild 3.2
.Sh NAME
.Nm mkify
.Nd BSDBuild auto-bundling script
.Sh SYNOPSIS
.Nm
.Ar module
.Ar ...
.Sh DESCRIPTION
The
.Nm
utility copies installed BSDBuild include files from their system location
to
.Pa mk
under the working directory.
It accepts a list of BSDBuild modules as arguments (e.g.,
.Sq prog
will cause
.Xr build.prog.mk 5
to be installed).
.Nm
also scans for include directives and will automatically install all
dependencies as well.
.Pp
The BSDBuild build system is designed to be bundled with applications as
a stand-alone set of
.Xr make 1
include files, which are typically installed in a directory named
.Pa mk
in the top-level directory (or some other central location) in an
application source code, such that people building applications which
use the BSDBuild build system do not need to have BSDBuild or any
other software installed on their systems.
.Sh SEE ALSO
.Xr mkconfigure 1 ,
.Xr build.common.mk 5 ,
.Xr build.lib.mk 5 ,
.Xr build.man.mk 5 ,
.Xr build.prog.mk 5 ,
.Xr build.www.mk 5
.Pp
.Lk https://bsdbuild.hypertriton.com/


CHANGELOG.md
# Changelog

All notable changes to BSDBuild will be documented in this file. This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [3.2] - 2024-03-08

### Added

- Add support for [cmake](https://cmake.org). Add `--output-cmake` option to [**mkconfigure**](https://bsdbuild.hypertriton.com/man1/mkconfigure) to output macros usable from cmake.
- Make the `depend` target use `CXX` and `CXXFLAGS` when compiling C++.
- Provide a built-in `configure` target for regenerating `configure` from `configure.in`. New make variable `${MKCONFIGURE_FLAGS}` for passing arguments to `mkconfigure`.
- [**mkconfigure**](https://bsdbuild.hypertriton.com/man1/mkconfigure): New C++ specific directives: `cxx_define()`, `cxx_incdir()`, `cxx_option()`, `check_cxx_header()`, `check_cxx_header_opts()` and `check_cxx_func_opts()`.
- [**mkconfigure**](https://bsdbuild.hypertriton.com/man1/mkconfigure): New directives: `ada_option()`, `ada_bflag()` and `hdefine_if()`.
- [**mkconfigure**](https://bsdbuild.hypertriton.com/man1/mkconfigure): New `pkgconfig_module()` directive for integrated pkg-config module generation. Add `$PKGCONFIG_LIBDIR` setting to simplify the installation of .pc modules.
- [**mkconfigure**](https://bsdbuild.hypertriton.com/man1/mkconfigure): Introduce `--keep-conftest` configure option (preserve output test files).
- [**build.prog.mk**](https://bsdbuild.hypertriton.com/man5/build.prog.mk) & [**build.lib.mk**](https://bsdbuild.hypertriton.com/man5/build.lib.mk): Add support for the Ada language.
- [**build.prog.mk**](https://bsdbuild.hypertriton.com/man5/build.prog.mk) & [**build.lib.mk**](https://bsdbuild.hypertriton.com/man5/build.lib.mk): Add support for compiling to wasm (WebAssembly) with Emscripten (https://emscripten.org).
- [**build.prog.mk**](https://bsdbuild.hypertriton.com/man5/build.prog.mk) & [**build.lib.mk**](https://bsdbuild.hypertriton.com/man5/build.lib.mk): Autodetect `${LINKER_TYPE}` default based on `${SRCS}` file extensions.
- [**build.prog.mk**](https://bsdbuild.hypertriton.com/man5/build.prog.mk) & [**build.lib.mk**](https://bsdbuild.hypertriton.com/man5/build.lib.mk): Introduce `${PROG_BUNDLE}`, `${LIB_BUNDLE}` and `gen-bundle.pl` for generating platform-specific application bundles.
- [**build.www.mk**](https://bsdbuild.hypertriton.com/man5/build.www.mk): Introduce the `mlproc(1)` multilanguage preprocessor. Define `${MLPROC}`, `${MLPROCFLAGS}`, `${MINIFIER}`, `${MINIFIERFLAGS}` and `${MINIFIERFLAGSCSS}`.
- **ada**: New test for Ada toolchain.
- **agar-ada**: New test module for Ada bindings to Agar-GUI.
- **agar-ada-core**: New test module for Ada bindings to Agar-Core.
- **agar.defs**, **agar-core.difs**: New test modules to generate tables of definitions and struct sizes for an installed Agar build. Intended to simplify the implementation of thin and variable-thickness bindings to different languages.
- **cc**: Add support for [cc65](https://cc65.github.io). Add `$CC_COMPILE`. Don't assume the compiler understands "-c" or "-O2".
- **db5**: New test for DB5.
- **fdclose**: New test for the `fdclose()` function.
- **devkitpro**: New test for [devkitPro](https://devkitpro.org/).
- **gle**: New test for [GLE](https://linas.org/gle).
- **imagemagick**: New test for ImageMagick 6 and 7.
- **libircclient**: New test for [libircclient](http://www.ulduzsoft.com/libircclient).
- **lldb**: New test for [LLDB](https://lldb.llvm.org/) and its utility library.
- **mprotect**: New test for the `mprotect()` call.
- **pcre**: New test for the [PCRE](https://pcre.org) library.
- **sdl2**: New test for [SDL 2.0](https://libsdl.org).
- **sdl_cpuinfo**: New test for the SDL cpuinfo functions.
- **sys_stat**: New test for `<sys/stat.h>`.
- **tcl**: New test for [TCL](https://tcl.tk/). Thanks Chuck!
- **xbox**: New test for Xbox XDK.
- **zlib**: New test for zlib.

### Removed

- Removed **build.proj.mk** module. Premake is no longer maintained and project file generation can be done with cmake.
- [**build.man.mk**](https://bsdbuild.hypertriton.com/man5/build.man.mk): Removed support for catman (`.cat*`) file auto-generation.
- [**mkconfigure**](https://bsdbuild.hypertriton.com/man1/mkconfigure): Removed unused Core function `Which()`.
- Removed **h2mandoc.pl**, **cmpfiles.pl** and **cleanfiles.pl**.
- Removed modules **md5**, **rmd160**, **sha1**.
- Removed module **freesg** (replaced by **agar-sg**).
- Removed module **freesg-rg** (replaced by **agar-map**).

### Changed

- Updates to the manual pages.
- Update config.guess to 2015-03-04. Add [FabBSD](https://FabBSD.org).
- In "install" target, create `${DESTDIR}` if needed.
- In **mkconcurrent**, extract and cache make target fragments directly from an installed build.\*.mk so we no longer need to maintain a copy of them in the script.
- [**build.lib.mk**](https://bsdbuild.hypertriton.com/man5/build.lib.mk): Disable use of Libtool by default. Generate .la files even when `USE_LIBTOOL=No`. Handle .dylib and .dll files directly.
- [**build.prog.mk**](https://bsdbuild.hypertriton.com/man5/build.prog.mk) & [**build.lib.mk**](https://bsdbuild.hypertriton.com/man5/build.lib.mk): Use ``{LIB,PROG}_PROFILE`` instead of separate .po targets. Merge the "depend" from [**build.dep.mk**](https://bsdbuild.hypertriton.com/man5/build.dep.mk). Include support for cc65 compiler.
- [**mkconfigure**](https://bsdbuild.hypertriton.com/man1/mkconfigure): Make the generated config.log a runnable shell script containing all test code and compilation commands (so individual tests can be reproduced easily by copy/pasting from it). Fix quoting of arguments in config.status.

mkify.pl
#!%PERL%
#
# Copyright (c) 2001-2020 Julien Nadeau Carriere <vedge@csoft.net>
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
# 1. Redistributions of source code must retain the above copyright
#    notice, this list of conditions and the following disclaimer.
# 2. Redistributions in binary form must reproduce the above copyright
#    notice, this list of conditions and the following disclaimer in the
#    documentation and/or other materials provided with the distribution.
# 
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE FOR
# ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
# DAMAGES (INCLUDING BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
# USE OF THIS SOFTWARE EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
#

@Scripts = qw(
	config.guess
	mkdep
	mkdep-cxx
	mkconcurrent.pl
	manlinks.pl
	gen-bundle.pl
	gen-declspecs.pl
	gen-dotdepend.pl
	gen-includes.pl
	gen-includelinks.pl
	gen-revision.sh
	gen-wwwdepend.pl);

@LibtoolFiles = qw(
	Makefile
	Makefile.in
	aclocal.m4
	config.guess
	config.sub
	configure
	configure.in
	install-sh
	ltmain.sh
	README);
@LibtoolFilesM4 = qw(
	libtool.m4
	ltoptions.m4
	ltsugar.m4
	ltversion.m4
	lt~obsolete.m4);

$DATADIR = '%DATADIR%';
%copied = ();

#
# Install file as-is.
# 
sub InstallASIS
{
	my $file = shift;
	my $dst = shift;
	
	if (!defined($dst)) { $dst = 'mk/'.$file; }

	print STDERR " $file";
	unless (open(SRC, $DATADIR.'/'.$file)) {
		print STDERR "Reading $DATADIR/$file: $!\n";
		return (0);
	}
	unless (open(DST, ">$dst")) {
		print STDERR "Writing to $dst: $!\n";
		return (0);
	}
	print DST <SRC>;
	close(DST);
	close(SRC);
	return (1);
}

#
# Copy files needed by <build.www.mk>.
#
sub CopyDepsWWW
{
	if (!-e 'xsl') {
		unless (mkdir('xsl', 0755)) {
			print STDERR "xsl/: $!\n";
			return (0);
		}
	}
	InstallASIS('ml.xsl', 'xsl/ml.xsl') || return (0);
}

#
# Copy files needed by <build.lib.mk>.
#
sub CopyDepsLIB
{
	if (! -e 'mk/libtool' && !mkdir('mk/libtool', 0755)) {
		print STDERR "mk/libtool/: $!\n";
		return (0);
	}
	if (! -e 'mk/libtool/m4' && !mkdir('mk/libtool/m4', 0755)) {
		print STDERR "mk/libtool/m4/: $!\n";
		return (0);
	}
	foreach my $f (@LibtoolFiles) {
		InstallASIS('libtool/'.$f) || return (0);
	}
	foreach my $f (@LibtoolFilesM4) {
		InstallASIS('libtool/m4/'.$f) || return (0);
	}
	chmod(0755, 'mk/libtool/config.sub') ||
	    print STDERR "mk/libtool/config.sub: $!\n";

	return (1);
}

#
# Install a .mk file. Scan the file for include dependency and install
# these as well.
#
sub InstallMK ($)
{
	my ($file) = @_;
	my $dst = 'mk/'.$file;
	my $path = $DATADIR.'/'.$file;
	my @deps = ();
	
	print STDERR " $file";

	unless (open(SRC, $DATADIR.'/'.$file)) {
		print STDERR "$DATADIR/$file: $!\n";
		return 0;
	}
	chop(@src = <SRC>);
	close(SRC);

	unless (open(DEST, ">$dst")) {
		print STDERR "dst $dst: $!\n";
		close(SRC);
		return 0;
	}
	foreach $_ (@src) {
		print DEST $_, "\n";

		if (/^include\s*.+\/mk\/(.+)\s*$/) {
			if (!exists($copied{$1})) {
				push @deps, $1;
				$copied{$1}++;
			}
		}
	}
	close(DEST);

	foreach my $dep (@deps) {
		InstallMK($dep);
	}
	if ($file eq 'build.www.mk') {
		CopyDepsWWW() || return (0);
	}
	if ($file eq 'build.lib.mk') {
		CopyDepsLIB() || return (0);
	}
	
	$copied{$file}++;
	return 1;
}

if (!-d 'mk' && @ARGV < 1) {
	print STDERR "Usage: $0 [module-name ...]\n";
	exit(1);
}

if (!-e 'mk' || @ARGV > 0) {
	if (! -e 'mk') {
		unless (mkdir('mk', 0755)) {
			print STDERR "mk/: $!\n";
			exit (1);
		}
	}
	print STDERR "Installing in mk:";
	foreach my $type (@ARGV) {
		InstallMK("build.${type}.mk");
	}
	print STDERR ".\n";
} else {
	unless (opendir(MKDIR, 'mk')) {
		print "mk/: $!\n";
		exit(1);
	}
	print STDERR "Updating mk:";
	foreach my $f (readdir(MKDIR)) {
		if ($f =~ /^\./ || -d 'mk/'.$f) { next; }
		if ($f =~ /^(build\.[\w]+\.mk)$/) {
			InstallMK($f);
		} elsif (-e $DATADIR.'/'.$f) {
			InstallASIS($f);
		}
	}
	print STDERR ".\n";
	closedir(MKDIR);
}

print STDERR "Installing scripts:";
foreach my $script (@Scripts) {
	InstallASIS($script);
}
print STDERR ".\n";

if (!-e 'Makefile' &&
    open(MAKE, '>Makefile')) {
	if ($type eq 'prog') {
		print MAKE << 'EOF';
# Sample Makefile for a program.

# Path to parent directory of "mk".
TOP=.

# Executable output name (exact meaning is platform-dependent)
PROG=foo

#
# Source files. See the <build.prog.mk> source for a list of all
# possible source types allowed.
# 
SRCS=foo.c bar.cc baz.m

include ${TOP}/Makefile.config
include ${TOP}/mk/build.prog.mk
EOF
	} elsif ($type eq 'lib') {
		print MAKE << 'EOF';
# Sample Makefile for a library.

# Path to parent directory of "mk".
TOP=.

# Library output name (exact meaning is platform-dependent)
LIB=	foo

#
# Source files. See the <build.lib.mk> source for a list of all
# possible source types allowed.
# 
SRCS=	foo.c bar.cc baz.m

include ${TOP}/Makefile.config
include ${TOP}/mk/build.lib.mk
EOF
	} elsif ($type eq 'www') {
		print MAKE << 'EOF';
# Sample Makefile for a webpage or website.

# Path to parent directory of "mk".
TOP=.

#
# Target HTML files. The actual source files must be named foo.htm and
# bar.htm, and will be processed into a number of language and character
# set variants (ie. foo.html.en). Standard Apache-compatible variant maps
# will be generated as well.
# 
HTML=	foo.html bar.html

include ${TOP}/mk/build.www.mk
EOF
	}
	close(MAKE);
}
if (!-e 'Makefile.config' &&
    open(MAKE, '>Makefile.config')) {
	print MAKE "# File is auto-generated, do not edit\n";
	close(MAKE);
}
